### PR TITLE
Enhancement in Trie Container

### PR DESCRIPTION
**Pull request by VivekReddy98**
Methods added in Trie:
1) remove 
2) size (Number of Legit words in the Container)

Tests & Examples Written for all the Supported Methods in Trie
1) insert
2) search
3) remove
4) size

### CODE REVIEW

File: include/forest/Trie.hpp [hunk 1]
```diff
@@ -1,29 +1,61 @@
...
 public:
+
+  Trie() = default; // Default Constructor;
+
+  /**
+   * Number of Legit Words in the Trie
+   */ 
+  int size(){
+    return this->numWords;
+  }
```
Comment (xorz57) [inline include/forest/Trie.hpp:35]: Maybe we should consider using a larger type than `int` here.

File: include/forest/Trie.hpp [hunk 2]
```diff
@@ -1,29 +1,61 @@
...
+
+  struct TrieNode { // TrieNode  
+    std::unordered_map<T, std::shared_ptr<TrieNode>> children;
+    bool end = false; // If True, prefix ending with this Node is a word.
   };
-  std::shared_ptr<Node> root = std::make_shared<Node>();
+
+  std::shared_ptr<TrieNode> root = std::make_shared<TrieNode>();
+
+  int numWords = 0;
```
Comment (xorz57) [inline include/forest/Trie.hpp:24]: Maybe we should consider using a larger type than `int` here.

File: examples/src/ExampleTrie1.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    Trie.insert("υπολογιστή");
+    Trie.insert("компьютер");
+
+    cout << Trie.size() << endl;
+    cout << Trie.search("కంప్యూటర్")  << endl;
+    cout << Trie.remove("комп")  << endl;
+
+    Trie.insert("Computer Science");
+    cout << Trie.size() << endl;
+    cout << Trie.search("संगणक")  << endl;
```
Comment (xorz57) [inline examples/src/ExampleTrie1.cpp:22]: https://docs.microsoft.com/en-us/windows/win32/learnwin32/working-with-strings

I am not 100% sure, but I think we should put these lines in a different example because it won't work on Microsoft Visual C++ compiler. Again, I am not 100% sure.

Comment (xorz57) [reply examples/src/ExampleTrie1.cpp:22]: We also have to make a third example file with the `wchar_t` type

File: include/forest/Trie.hpp [hunk 4]
```diff
@@ -1,29 +1,61 @@
...
 
-template <typename T> class trie {
+namespace forest{
+template <typename T> class Trie {
 private:
-  struct Node {
-    std::unordered_map<T, std::shared_ptr<Node>> children;
-    bool end = false;
+
+  struct TrieNode { // TrieNode
```
Comment (xorz57) [inline include/forest/Trie.hpp:17]: I chose the name `Node` instead of `TrieNode` because the user won't have access to it anyway because it is defined as private.

File: include/forest/Trie.hpp [hunk 5]
```diff
@@ -33,4 +65,43 @@ template <typename T> class trie {
...
+
+  /**
+   * Iterative Implementation of remove using a Stack.
+   * Returns True if the String is Found & Deletes it.
+   *         False Otherwise
+   */
+  bool remove(const std::basic_string<T> &key) {
+    std::shared_ptr<TrieNode> current = root;
+    std::stack<std::shared_ptr<TrieNode>> stk;
+
```
Comment (xorz57) [inline include/forest/Trie.hpp:77]: Return early if `current` is `nullptr` (in case of a completely empty Trie)
```cpp
    if (!current)
      return false;
```

Comment (xorz57) [reply include/forest/Trie.hpp:77]: Maybe this should also be placed before creating the stack.

File: include/forest/Trie.hpp [hunk 6]
```diff
@@ -33,4 +65,43 @@ template <typename T> class trie {
...
+    for(const T &c : key){
+        if (current->children.empty())
+            return false;
+        current = current->children[c];
+        if (!current)
+          return false;
+        stk.push(current); // Push all the Node Pointers if Present
+    }
+
+    if (stk.size() <= 1 || !stk.top()->end) return false;
```
Comment (xorz57) [inline include/forest/Trie.hpp:89]: If we check if `current` is `nullptr` before the loop would `Line 89` still be necessary? I got confused a little bit.

File: tests/src/TestTrie.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+#include <forest/Trie.hpp>
+
+SCENARIO("Test Insert/Search methods in Trie Data Structure") {
+  GIVEN("An empty Trie") {
+    forest::Trie<char16_t> trie;
+    REQUIRE(trie.search(u"computer") == false);
+    REQUIRE(trie.search(u" ") == false);
+    REQUIRE(trie.search(u" Science ") == false);
+    REQUIRE(trie.size() == 0);
+
```
Comment (xorz57) [inline tests/src/TestTrie.cpp:12]: Try removing words from an empty Trie.
We also need to try the empty string `""` and `u""`.

File: tests/src/TestTrie.cpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
+#define CATCH_CONFIG_MAIN
+#include <catch2/catch.hpp>
+#include <forest/Trie.hpp>
+
+SCENARIO("Test Insert/Search methods in Trie Data Structure") {
+  GIVEN("An empty Trie") {
+    forest::Trie<char16_t> trie;
+    REQUIRE(trie.search(u"computer") == false);
+    REQUIRE(trie.search(u" ") == false);
```
Comment (xorz57) [inline tests/src/TestTrie.cpp:9]: Try the empty string `""` and `u""` as a key also.

File: tests/src/TestTrie.cpp [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+SCENARIO("Test Remove method in Trie Data Structure") {
+
+    forest::Trie<char16_t> trie;
+    REQUIRE(trie.size() == 0);
+    trie.insert(u"Chemical");
+    trie.insert(u"Computer");
+    trie.insert(u"Computer Science");
+    trie.insert(u"Computer Engineering");
+
```
Comment (xorz57) [inline tests/src/TestTrie.cpp:46]: Try removing a word when the Trie is empty.

Comment (xorz57) [reply tests/src/TestTrie.cpp:46]: Also, try the empty string (`""` and `u""`) as a key

**Comment by xorz57** [comment]
image

No memory leaks! 👍

File: include/forest/Trie.hpp (hunk 1 above)
Comment (xorz57) [reply include/forest/Trie.hpp:35]: We can change that after merging this pull request. It is OK!

File: include/forest/Trie.hpp (hunk 2 above)
Comment (xorz57) [reply include/forest/Trie.hpp:24]: We can change that after merging this pull request. It is OK!

File: include/forest/Trie.hpp (hunk 1 above)
Comment (VivekReddy98) [reply include/forest/Trie.hpp]: Made numWords long long, for now, can change it later if needed.

File: include/forest/Trie.hpp (hunk 5 above)
Comment (VivekReddy98) [reply include/forest/Trie.hpp:83]: The right way is to check if root is empty would be (!root->children.empty()). I've added this check on the top of the method, just not to bother allocating resources even when Trie is empty. It would still be caught at 
`if (stk.size() <= 1 || !stk.top()->end) return false;`

File: include/forest/Trie.hpp (hunk 6 above)
Comment (VivekReddy98) [reply include/forest/Trie.hpp:96]: This is required to check If No Prefix is matched or The prefix is not a Legit word. So this cannot be reviewed

File: tests/src/TestTrie.cpp (hunk 8 above)
Comment (VivekReddy98) [reply tests/src/TestTrie.cpp:45]: Thank you so much for pointing out, I almost forgot this corner case. It's fixed now.

File: tests/src/TestTrie.cpp (hunk 9 above)
Comment (VivekReddy98) [reply tests/src/TestTrie.cpp:82]: Empty Trie Tests Added

File: examples/src/ExampleTrie1.cpp (hunk 3 above)
Comment (VivekReddy98) [reply examples/src/ExampleTrie1.cpp]: I've added different examples for char, char16_t & wchar_t. However, I was not able to test this on the MSVC compiler. I was not able to build the project on Windows infact. If you have any instructions for building on Visual Studio, please let me know.

File: include/forest/Trie.hpp [hunk 10]
```diff
@@ -21,7 +22,7 @@ template <typename T> class Trie {
 
   std::shared_ptr<TrieNode> root = std::make_shared<TrieNode>();
 
-  int numWords = 0;
+  long long numWords = 0;
```
Comment (xorz57) [inline include/forest/Trie.hpp:25]: That or `std::size_t`.

File: include/forest/Trie.hpp (hunk 5 above)
Comment (xorz57) [reply include/forest/Trie.hpp:83]: Sounds good!

**Comment by xorz57** [comment]
@allcontributors please add @VivekReddy98 for code and tests.

**Comment by xorz57** [comment]
@all-contributors please add @VivekReddy98 for code and tests.

**Comment by xorz57** [comment]
@all-contributors please add @VivekReddy98 for code and tests.

**Comment by xorz57** [comment]
@all-contributors please add @VivekReddy98 for code and tests.
